gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/GradleCacheExplorer.java

```diff
@@ -0,0 +1,155 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.versions.intellij;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.progress.ProgressIndicator;
+import com.intellij.openapi.progress.ProgressManager;
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GradleCacheExplorer {
+
+    private static final Logger log = LoggerFactory.getLogger(GradleCacheExplorer.class);
+    private static final String GRADLE_CACHE_PATH = System.getProperty("user.home") + "/.gradle/caches/modules-2/";
+    private static final Cache<String, Set<String>> cache =
+            Caffeine.newBuilder().expireAfterWrite(10, TimeUnit.MINUTES).build();
+
+    private final List<String> projectUrls;
+
+    public GradleCacheExplorer(List<String> projectUrls) {
+        this.projectUrls = projectUrls;
+    }
+
+    public final Set<String> getCompletions(DependencyGroup input) {
+        String parsedInput = String.join(".", input.parts());
+        ProgressIndicator indicator = ProgressManager.getInstance().getProgressIndicator();
+        if (indicator == null) {
+            return new HashSet<>();
+        }
+
+        try {
+            Future<Set<String>> future =
+                    ApplicationManager.getApplication().executeOnPooledThread(extractStrings(indicator));
+            Set<String> results =
+                    com.intellij.openapi.application.ex.ApplicationUtil.runWithCheckCanceled(future::get, indicator);
+
+            return results.stream()
+                    .filter(result -> result.startsWith(parsedInput))
+                    .map(result -> parsedInput.isEmpty() ? result : result.substring(parsedInput.length() + 1))
+                    .collect(Collectors.toSet());
+        } catch (RuntimeException e) {
+            log.debug("Operation was cancelled", e);
+        } catch (Exception e) {
+            log.warn("Failed to get completions", e);
+        }
+        return new HashSet<>();
+    }
+
+    private Callable<Set<String>> extractStrings(ProgressIndicator indicator) {
+        return () -> cache.get("metadata", key -> {
+            Set<String> stringsSet = new HashSet<>();
+            File gradleCacheFolder = new File(GRADLE_CACHE_PATH);
+            File[] metadataFolders = gradleCacheFolder.listFiles((dir, name) -> name.startsWith("metadata-"));
+
+            if (metadataFolders != null) {
```

Comment:
I've updated this to use a Files so that it can all be a stream I think this is much cleaner is this what you meant?

``` java
    private Set<String> extractStrings(ProgressIndicator indicator) {
        return cache.get("metadata", key -> {
            try (Stream<Path> metadataFolders = Files.list(Paths.get(GRADLE_CACHE_PATH))
                    .filter(path -> path.getFileName().toString().startsWith("metadata-"))) {
                return metadataFolders
                        .filter(metadataFolder -> !indicator.isCanceled())
                        .map(metadataFolder -> metadataFolder.resolve("resource-at-url.bin"))
                        .filter(Files::exists)
                        .flatMap(this::extractStringsFromBinFile)
                        .filter(this::isValidResourceUrl)
                        .map(this::extractGroupAndArtifactFromUrl)
                        .flatMap(Optional::stream)
                        .collect(Collectors.toSet());
            } catch (IOException e) {
                log.error("Failed to list metadata folders", e);
                return Collections.emptySet();
            }
        });
    }
```

Reply:
After our chat yesterday I have updated this to throw the exception properly